pydatastructs/linear_data_structures/arrays.py

```diff
@@ -133,6 +135,111 @@ def __len__(self):
+
+    Parameters
+    ==========
+
+    dtype: type
+        A valid object type.
+    *args: int
+        The dimensions of the array.
+
+    Raises
+    ======
+
+    IndexError
+        Index goes out of boundaries, or
+        the number of index given is not
+        the same as the number of dimensions.
+    ValueError
+        When there's no dimensions or the
+        dimension size is 0.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import MultiDimensionalArray as MDA
+    >>> arr = MDA(int, 5, 6, 9)
+    >>> arr.fill(32)
+    >>> arr[3, 0, 0]
+    32
+    >>> arr[3, 0, 0] = 7
+    >>> arr[3, 0, 0]
+    7
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Array_data_structure#Multidimensional_arrays
+
+    """
+    __slots__ = ['_sizes', '_data', '_dtype']
+
+    def __new__(cls, dtype=NoneType, *args, **kwargs):
+        if dtype is NoneType or len(args) == (0):
+            raise ValueError("Array cannot be created due to incorrect"
+                             " information.")
+        if len(args) == 1:
+            obj = Array.__new__(cls)
+            obj._dtype = dtype
+            obj._shape = [1]
+            obj._data = [None] * args[0]
+            return obj
+        dimensions = args
+        for dimension in dimensions:
+            if dimension < 1:
+                raise ValueError("Array cannot be created due to incorrect"
+                                 " dimensions.")
+        n_dimensions = len(dimensions)
+        d_sizes = []
+        index = 0
+        while n_dimensions > 1:
+            size = dimensions[index]
+            for i in range(index,  len(dimensions) - 1):
+                size = size * dimension
+            d_sizes.append(size)
+            n_dimensions -= 1
+            index += 1
+        d_sizes.append(dimensions[index])
+        d_sizes.append(1)
+        obj = Array.__new__(cls)
+        obj._dtype = dtype
+        obj._shape = d_sizes
+        obj._data = [None] * obj._shape[1] * dimensions[0]
+        return obj
+
+    def __getitem__(self, indices):
+        if type(indices) == int:
+            if len(self._shape) > 1:
+                raise IndexError("Shape mismatch, current shape is %s" % (self._shape))
+            else:
+                return self._data[indices]
+        if len(indices) != len(self._shape) - 1:
+            raise IndexError("Shape mismatch, current shape is %s"%(self._shape))
+        position = 0
+        for i in range(0, len(indices)):
+            position += self._shape[i+1] * indices[i]
+        return self._data[position]
+
+    def __setitem__(self, indices, element):
+        if type(indices) == int:
+            self._data[indices] = element
+        elif len(indices) != len(self._shape) - 1:
+            raise IndexError("Number of dimensions not expected.")
+        else:
+            position = 0
+            for i in range(0, len(indices)):
+                position += self._shape[i+1] * indices[i]
```

Comment:
Probably, the check for correct indices can be done in the following way,

1. Check if size of `self._shape` and `indices` is same.
2. Then, check if each indices[i] < self._shape[i]. 

There can be a private function(name starting with, `_`) which can do the check, avoiding repetition in `__getitem__` and `__setitem__`.